File: smoke_tests/swift/file_upload.swift
// Insecure File Upload vulnerabilities in Swift
import Foundation

class FileUploadVulnerabilities {

    let uploadDir = "/var/uploads"

    // Test 1: No file type validation
    func uploadFile(filename: String, data: Data) -> Bool {
        // VULNERABLE: No file type check
        let path = "\(uploadDir)/\(filename)"
        return FileManager.default.createFile(atPath: path, contents: data)
    }

    // Test 2: Client-provided content type
    func uploadWithContentType(filename: String, data: Data, contentType: String) -> Bool {
        // VULNERABLE: Trusting client content type
        if contentType.hasPrefix("image/") {
            let path = "\(uploadDir)/\(filename)"
            return FileManager.default.createFile(atPath: path, contents: data)
        }
        return false
    }

    // Test 3: Extension-only validation
    func uploadImage(filename: String, data: Data) -> Bool {
        let validExtensions = ["jpg", "jpeg", "png", "gif"]
        let ext = (filename as NSString).pathExtension.lowercased()

        // VULNERABLE: Only checking extension, not content
        if validExtensions.contains(ext) {
            let path = "\(uploadDir)/\(filename)"
            return FileManager.default.createFile(atPath: path, contents: data)
        }
        return false
    }

    // Test 4: Executable upload
    func uploadScript(filename: String, data: Data) -> Bool {
        // VULNERABLE: Allows executable uploads
        let path = "\(uploadDir)/\(filename)"
        FileManager.default.createFile(atPath: path, contents: data)
        // Make executable
        try? FileManager.default.setAttributes(
            [.posixPermissions: 0o755],
            ofItemAtPath: path
        )
        return true
    }

    // Test 5: No size limit
    func uploadLargeFile(filename: String, data: Data) -> Bool {
        // VULNERABLE: No file size check (DoS risk)
        let path = "\(uploadDir)/\(filename)"
        return FileManager.default.createFile(atPath: path, contents: data)
    }

    // Test 6: Path traversal in filename
    func uploadDocument(filename: String, data: Data) -> URL? {
        // VULNERABLE: No sanitization of filename
        let path = "\(uploadDir)/\(filename)"
        if FileManager.default.createFile(atPath: path, contents: data) {
            return URL(fileURLWithPath: path)
        }
        return nil
    }

    // Test 7: Double extension bypass
    func uploadSafeFile(filename: String, data: Data) -> Bool {
        // VULNERABLE: Can bypass with file.php.jpg
        let ext = (filename as NSString).pathExtension.lowercased()
        let dangerousExtensions = ["php", "exe", "sh", "bat"]

        if dangerousExtensions.contains(ext) {
            return false
        }
        let path = "\(uploadDir)/\(filename)"
        return FileManager.default.createFile(atPath: path, contents: data)
    }

    // Test 8: Null byte injection
    func uploadWithNullByte(filename: String, data: Data) -> Bool {
        // VULNERABLE: Null byte in filename
        let path = "\(uploadDir)/\(filename)"
        return FileManager.default.createFile(atPath: path, contents: data)
    }

    // Test 9: SVG upload (XSS vector)
    func uploadSvg(filename: String, data: Data) -> Bool {
        // VULNERABLE: SVG can contain JavaScript
        if filename.hasSuffix(".svg") {
            let path = "\(uploadDir)/\(filename)"
            return FileManager.default.createFile(atPath: path, contents: data)
        }
        return false
    }

    // Test 10: Archive upload without scanning
    func uploadArchive(filename: String, data: Data) -> Bool {
        let archiveExtensions = ["zip", "tar", "gz"]
        let ext = (filename as NSString).pathExtension.lowercased()

        // VULNERABLE: No content scanning of archives
        if archiveExtensions.contains(ext) {
            let path = "\(uploadDir)/\(filename)"
            return FileManager.default.createFile(atPath: path, contents: data)
        }
        return false
    }

    // Test 11: Upload to web-accessible directory
    func uploadToWebroot(filename: String, data: Data) -> Bool {
        // VULNERABLE: Uploading to web-accessible location
        let webrootPath = "/var/www/html/uploads/\(filename)"
        return FileManager.default.createFile(atPath: webrootPath, contents: data)
    }

    // Test 12: Image with embedded PHP
    func uploadImageFile(filename: String, data: Data) -> Bool {
        // VULNERABLE: Image magic bytes but embedded code
        let header = data.prefix(8)
        let pngMagic = Data([0x89, 0x50, 0x4E, 0x47])
        let jpgMagic = Data([0xFF, 0xD8, 0xFF])

        if header.starts(with: pngMagic) || header.starts(with: jpgMagic) {
            // Only checking header, not full content
            let path = "\(uploadDir)/\(filename)"
            return FileManager.default.createFile(atPath: path, contents: data)
        }
        return false
    }

    // Test 13: Temporary file not cleaned up
    func processUpload(filename: String, data: Data) -> Bool {
        let tempPath = "/tmp/\(UUID().uuidString)_\(filename)"
        // VULNERABLE: Temp file left on disk
        FileManager.default.createFile(atPath: tempPath, contents: data)
        // Process file but never delete temp
        return processFile(path: tempPath)
    }

    private func processFile(path: String) -> Bool { true }
}

File: smoke_tests/dart/file_upload.dart
// File Upload vulnerabilities in Dart

import 'dart:io';
import 'dart:convert';

// Test 1: No file type validation
Future<void> vulnerableNoTypeValidation(HttpRequest request) async {
  // VULNERABLE: Accepting any file type
  var boundary = request.headers.contentType?.parameters['boundary'];
  var transformer = MimeMultipartTransformer(boundary!);
  await for (var part in transformer.bind(request)) {
    var filename = getFilename(part.headers);
    var file = File('/uploads/$filename');
    await part.pipe(file.openWrite());
  }
}

// Test 2: Client-side filename used directly
Future<void> vulnerableUnsafeFilename(String clientFilename, List<int> data) async {
  // VULNERABLE: Using client-provided filename without sanitization
  var file = File('/uploads/$clientFilename');
  await file.writeAsBytes(data);
}

// Test 3: Path traversal in upload
Future<void> vulnerablePathTraversalUpload(String filename, List<int> content) async {
  // VULNERABLE: Filename may contain ../
  var uploadPath = '/var/www/uploads/$filename';
  var file = File(uploadPath);
  await file.writeAsBytes(content);
}

// Test 4: Executable file upload
Future<void> vulnerableExecutableUpload(HttpRequest request) async {
  // VULNERABLE: No check for executable extensions
  var filename = request.headers.value('X-Filename')!;
  var content = await request.toList();
  var file = File('/uploads/$filename');
  await file.writeAsBytes(content.expand((x) => x).toList());
}

// Test 5: MIME type from client
Future<void> vulnerableTrustedMimeType(HttpRequest request) async {
  // VULNERABLE: Trusting client-provided content type
  var contentType = request.headers.contentType?.mimeType;
  if (contentType == 'image/jpeg' || contentType == 'image/png') {
    // Accepts file based on client header
    var data = await request.toList();
    await File('/uploads/image.dat').writeAsBytes(data.expand((x) => x).toList());
  }
}

// Test 6: No file size limit
Future<void> vulnerableNoSizeLimit(HttpRequest request) async {
  // VULNERABLE: No size limit on upload
  var data = <int>[];
  await for (var chunk in request) {
    data.addAll(chunk);
  }
  await File('/uploads/large_file.dat').writeAsBytes(data);
}

// Test 7: Double extension bypass
Future<void> vulnerableDoubleExtension(String filename, List<int> data) async {
  // VULNERABLE: Only checking last extension
  if (filename.endsWith('.jpg') || filename.endsWith('.png')) {
    // file.php.jpg would pass
    await File('/uploads/$filename').writeAsBytes(data);
  }
}

// Test 8: Null byte injection
Future<void> vulnerableNullByte(String filename, List<int> data) async {
  // VULNERABLE: Null byte could truncate extension check
  var uploadPath = '/uploads/$filename';
  await File(uploadPath).writeAsBytes(data);
}

// Test 9: Upload to web root
Future<void> vulnerableWebRootUpload(String filename, List<int> data) async {
  // VULNERABLE: Uploading to publicly accessible directory
  var webRoot = '/var/www/html';
  await File('$webRoot/uploads/$filename').writeAsBytes(data);
}

// Test 10: SVG upload (XSS vector)
Future<void> vulnerableSvgUpload(String svgContent) async {
  // VULNERABLE: SVG can contain JavaScript
  await File('/uploads/image.svg').writeAsString(svgContent);
}

// Test 11: ZIP file upload without scanning
Future<void> vulnerableZipUpload(List<int> zipData) async {
  // VULNERABLE: ZIP could contain malicious files
  await File('/uploads/archive.zip').writeAsBytes(zipData);
  // Extracts without validation
  await Process.run('unzip', ['/uploads/archive.zip', '-d', '/uploads/extracted']);
}

// Test 12: Overwrite existing files
Future<void> vulnerableFileOverwrite(String filename, List<int> data) async {
  // VULNERABLE: Can overwrite existing files
  var file = File('/uploads/$filename');
  await file.writeAsBytes(data); // No check if file exists
}

// Test 13: Race condition in upload
Future<void> vulnerableRaceCondition(String filename, List<int> data) async {
  // VULNERABLE: TOCTOU race condition
  var file = File('/uploads/$filename');
  if (!await file.exists()) {
    await Future.delayed(Duration(milliseconds: 10)); // Simulated delay
    await file.writeAsBytes(data);
  }
}

// Test 14: Symlink attack
Future<void> vulnerableSymlinkAttack(String filename, List<int> data) async {
  // VULNERABLE: Following symlinks
  var file = File('/uploads/$filename');
  await file.writeAsBytes(data); // May follow symlink to sensitive location
}

// Test 15: Image processing without validation
Future<void> vulnerableImageProcessing(List<int> imageData) async {
  // VULNERABLE: Processing potentially malicious image
  var tempFile = File('/tmp/temp_image.jpg');
  await tempFile.writeAsBytes(imageData);
  // ImageMagick could be vulnerable to image-based exploits
  await Process.run('convert', [tempFile.path, '-resize', '100x100', '/uploads/thumb.jpg']);
}

// Helper functions
String? getFilename(Map<String, String> headers) {
  var contentDisposition = headers['content-disposition'];
  if (contentDisposition != null) {
    var match = RegExp(r'filename="([^"]*)"').firstMatch(contentDisposition);
    return match?.group(1);
  }
  return null;
}

// Mock transformer
class MimeMultipartTransformer extends StreamTransformerBase<List<int>, MimeMultipart> {
  final String boundary;
  MimeMultipartTransformer(this.boundary);

  @override
  Stream<MimeMultipart> bind(Stream<List<int>> stream) {
    throw UnimplementedError();
  }
}

class MimeMultipart extends Stream<List<int>> {
  final Map<String, String> headers = {};

  @override
  StreamSubscription<List<int>> listen(void Function(List<int>)? onData, {Function? onError, void Function()? onDone, bool? cancelOnError}) {
    throw UnimplementedError();
  }
}
